lms/product/plugin/course_copy.py

```diff
@@ -0,0 +1,74 @@
+from lms.models import File
+from lms.services.exceptions import CanvasFileNotFoundInCourse
+
+
+class CourseCopyPlugin:
+    file_type = None
+
+    def assert_file_in_course(self, course_id, file_id):
+        """Raise if the current user can't see file_id in course_id."""
+        raise NotImplementedError()
+
+    def find_matching_file_in_course(self, original_file_id, new_course_id):
+        try:
+            # Get the current (copied) courses files, that will have the side effect of storing files in the DB
+            _ = self._store_new_course_files(new_course_id)
+        except ExternalRequestError:
+            # We might not have access to use the API for that endpoint.
+            # That will depend on our role and the course's permissions settings.
+            # We will continue anyway, maybe the files of the new course are already in the DB
+            # after an instructor launched corrected the issue.
+            pass
+
+        # We get the original file record from the DB
+        file = self._file_service.get(original_file_id, type_=self.file_type)
+        if not file:
+            # That file must have been recorded when the original assignment was configured.
+            # If we can't find that one something odd is going on, stop here.
+            return
+
+        # Now we'll try to find a matching file in the DB in the new course
+        # We might have a record of this because we just called `_store_new_course_files` as the current user
+        # or another user might have done it before for us.
+        new_file = (
+            self._db.query(File)
+            .filter(
+                # Same application instance, not entirely correct but fine (it should be same tool_guid, or ideally same org)
+                File.application_instance == file.application_instance,
+                # Looking for files in the original course only
+                File.course_id == new_course_id,
+                # Same type, `canvas_file` here
+                File.type == file.type,
+                # We don't want to find the same file we are looking for
+                File.lms_id != file.lms_id,
+                # And as a heuristic, we reckon same name, same size, probably the sme file
+                File.name == file.name,
+                File.size == file.size,
+            )
+            .first()
+        )
+
+        if new_file:
+            # We found the equivalent file in the new course
+            return new_file.lms_id
+
+        # No match for the file.
+        # This could be an issue with our heuristic to find the new file
+        # or other edge case, for example a file was deleted after course copy or similar.
+        return None
+
+    @staticmethod
+    def get_mapped_file_id(course, file_id):
+        """Get a previously mapped file id in course."""
+        return course.extra.get("course_copy_file_mappings", {}).get(file_id, file_id)
```

Comment:
Another change over the canvas solution, we store the mapping in the course not the assignment.

All of this course copy business is a problem because files belong to the course, we should map them the same way in our model.

This also has a theoretical benefit of allowing reusing the same file mapping for multiple assignments that use the same file.

Reply:
I think there was a reason why these mappings were stored in assignments rather than courses, but I can't remember.